### PR TITLE
Calculate connected components

### LINKED ISSUE #89: Connected components: 2. find connected components in `N`
related to: #86
focusing on 2. ii. in this issue.

the following functionality is needed:
given a list on `N` most recent transactions, return a list of clusters:
`[{tx_1, tx_2}, {tx_3}, {tx_4, tx_5, tx_6} ...]`

### PR DESCRIPTION
**Pull request by alongalky**
Fixes #89

### CODE REVIEW

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 1]
```diff
@@ -20,6 +21,53 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
...
+        Set<Hash> unvisited = new HashSet<>(transactions);
+        List<Set<Hash>> result = new ArrayList<>();
+
+        // Outer loop iterates once for each component
+        while (!unvisited.isEmpty()) {
+            Set<Hash> currentComponent = new HashSet<>();
+
+            // Perform DFS scan to find all elements in this component
+            Deque<Hash> stack = new ArrayDeque<>();
+            Hash currentHash = unvisited.iterator().next();
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: why do you need `currentHash` explicitly?
(other option to push directly to `stack`, and declare in internal loop)

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 2]
```diff
@@ -20,6 +21,53 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
...
+            stack.push(currentHash);
+            while (!stack.isEmpty()) {
+                currentHash = stack.pop();
+
+                if (!currentComponent.contains(currentHash)) {
+                    unvisited.remove(currentHash);
+                    currentComponent.add(currentHash);
+
+                    Collection<Hash> neighbors = getNeighbors(currentHash);
+                    Collection<Hash> unvisitedNeighbors = neighbors.stream()
```
Comment (alon-e) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: you don't actually need the set. so you can avoid collecting it and just pushing to the stack directly. using `forEach`.

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java (hunk 1 above)
Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: I wasn't too happy with the line:
```
stack.push(unvisited.iterator().next());
```

but I accept the change

**Review by alon-e — APPROVED** [review]

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 3]
```diff
@@ -20,6 +21,47 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
         this.random = new SecureRandom();
     }
 
+    public Collection<Set<Hash>> getConnectedComponents(Set<Hash> transactions) throws Exception {
+        Set<Hash> unvisited = new HashSet<>(transactions);
```
Comment (Wollac) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:25]: An alternative would be using `DAGHelper::traverseApprovees` (and thus BFS) to compute connected components and avoid code duplication.

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 4]
```diff
@@ -20,6 +21,47 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
...
+            stack.push(unvisited.iterator().next());
+            while (!stack.isEmpty()) {
+                Hash currentHash = stack.pop();
+
+                if (!currentComponent.contains(currentHash)) {
+                    unvisited.remove(currentHash);
+                    currentComponent.add(currentHash);
+
+                    getNeighbors(currentHash).stream()
+                        .filter(t -> unvisited.contains(t))
```
Comment (Wollac) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:43]: Replace with `filter(unvisited::contains)`

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 5]
```diff
@@ -20,6 +21,47 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
...
+                }
+            }
+
+            result.add(currentComponent);
+        }
+
+        return result;
+    }
+
+    private Collection<Hash> getNeighbors(Hash hash) throws Exception {
```
Comment (Wollac) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:54]: In graph theory these vertices are called _adjacent_, so maybe `getAdjacentTransactions` would be preferable especially as neighbors is already used in the IRI context.

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 6]
```diff
@@ -194,6 +194,59 @@ public void selectTipFromLargestConnectedComponentWithMultipleTips() throws Exce
         Assert.assertTrue(hairOnChainTransactions.contains(tip));
     }
 
+    @Test
+    public void getConnectedComponentsReturnsCorrectSetsForStar() throws Exception {
+        final int amount = 10;
+        Set<Hash> loneTransactions = new HashSet<Hash>(makeStar(amount, Hash.NULL_HASH, 0));
```
Comment (Wollac) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:200]: Prefer diamond operator for consistency.

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java (hunk 3 above)
Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:25]: The reason I didn't do that was that I want a bidirectional scan (towards both approvers and approvees), and DAGHelper doesn't provide that. I don't know whether the transaction I'm starting from in each component is a tip or somewhere in the middle.

**Pull request merged**